spec/Classes.tex

```diff
@@ -770,6 +1085,314 @@ \subsection{The Compiler-Generated Constructor}
+In the following code:
+\begin{chapel}
+class A {
+  var a, b : real;
+
+  proc init() {
+    this.init(1.0);
+  }
+
+  proc init(val : real) {
+    this.a = val;
+    this.b = val * 2;
+  }
+}
+
+class B : A {
+  var x, y : real;
+
+  proc init(val: real, x: real, y: real) {
+    super.init(val);
+    this.x = x;
+    this.y = y;
+  }
+
+  proc init() {
+    // implicit super.init();
+    this.x = a*2;
+    this.y = b*2;
+  }
+}
+
+var b1 = new B(4.0, 1.0, 2.0);
+var b2 = new B();
+\end{chapel}
+\begin{chapelpost}
+writeln(b1);
+writeln(b2);
+\end{chapelpost}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+\begin{chapeloutput}
+{a = 4.0, b = 8.0, x = 1.0, y = 2.0}
+{a = 1.0, b = 2.0, x = 2.0, y = 4.0}
+\end{chapeloutput}
+The first initializer explicitly calls an initializer for class \chpl{A}. Once
+the parent's initializer is complete, fields of class \chpl{B} may be
+initialized.
+
+The second initializer implicitly invokes the parent's initializer with zero
+arguments, and then uses the parent's fields to initialize its own fields.
+\end{chapelexample}
+
+As stated earlier, the compiler will insert a zero-argument call to the
+parent's initializer if the user has not explicitly written one themselves.
+The exception to this rule is if the initializer body invokes another
+initializer. This other initializer will either contain an implicit or explicit
+call to the parent initializer, and so the calling initializer should not
+attempt to initialize the parent itself. This also means that parent fields
+may not be accessed before explicit calls to \chpl{init}.
+
+\begin{chapelexample}{superInitThisInit.chpl}
+In the following code:
+\begin{chapel}
+class Parent {
+  var x, y: real;
+}
+
+class Child : Parent {
+  var z : real;
+
+  proc init(x: real, y: real, z: real) {
+    super.init(x, y);
+    this.z = z;
+  }
+
+  proc init(z: real) {
+    this.init(0.0, 0.0, z);
+  }
+}
+
+var c = new Child(5.0);
+\end{chapel}
+\begin{chapelpost}
+writeln(c);
+\end{chapelpost}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+\begin{chapeloutput}
+{x = 0.0, y = 0.0, z = 5.0}
+\end{chapeloutput}
+The second initializer does not contain an implicit call to the parent's
+initializer because it explicitly invokes another initializer.
+\end{chapelexample}
+
+\subsubsection{Calling Methods on Parent Classes}
+\label{Calling_Methods_on_Parent_Classes}
+\index{classes!initializers!initializer-inheritance!dynamic-this}
+
+Because parent classes are initialized before child classes, the
+partially-initialized memory can be treated as if it were the parent class
+once super.init() returns. The dynamic type of \chpl{this} is the parent's
+type until the \chpl{complete} method is invoked (except when local fields
+are initialized and used). This means that before the \chpl{complete} method
+is invoked, parent methods may be called on \chpl{this} and \chpl{this} may be
+passed to functions as the parent type.
```

Comment:
I thought above we were saying that can't happen... maybe check & add a cross reference from the part discussing complete to here.